### DISCUSSION TITLE
HNSW memory usage estimation

### DISCUSSION THREAD

**Discussion by rere950303** [milvus-io/milvus#33424] [Q&A and General discussion] [answered]
Can I expect memory usage if I index a 100K entity by hnsw? The dimension is 3072, so I'm going to make M about 48 of the hnsw build parameters. If M is too big, I'd appreciate it if you could advise me how much would be appropriate for same recall rate.

**Comment by yhmo** [comment, marked as answer]
HNSW size is a bit larger than original vectors size. Each dimension value is a float32 type, so each dimension value occupies 4 bytes.
original vectors size = 100K * 3072 * 4 Bytes = 1172MB
Then the HNSW size could be around 1200MB.

https://milvus.io/docs/index.md#HNSW
The range of M is (2, 2048). M=48 is ok, it is not a big value.

**Reply by rere950303** [reply]
@yhmo thank you for replying! 
I asked you a same question here(https://github.com/milvus-io/milvus/discussions/16484), the size of the memory required is not based on each querynode, but based on the sum of all querynode replications, right?

**Reply by yhmo** [reply]
1200MB is the total RAM size required.
Milvus manages data by segments. Typically, each segment max size is 512MB or 1024MB.
Assume original vectors size is 1172MB, and segment max size is 512MB, there will be 3 ~ 4 segments in the collection.
When loading the collection, milvus will try balance the data across query nodes.
If you have 2 query nodes, then each query node will load 2 segments. Each query node requires at least 600MB RAM.
If you have 5 query nodes, then there must be 1 query node is empty because the numer of segments is smaller than number of query nodes.

**Reply by rere950303** [reply]
Thank you. Perfectly understood!!
